pkg/controller/acmeorders/sync.go

```diff
@@ -700,32 +700,31 @@ func getACMEOrder(ctx context.Context, cl acmecl.Interface, o *cmacme.Order) (*a
 	return acmeOrder, nil
 }
 
-func getAltCertChain(ctx context.Context, cl acmecl.Interface, certURL string, preferredChain string) (bool, [][]byte, error) {
+func getPreferredCertChain(ctx context.Context, cl acmecl.Interface, certURL string, preferredChain string) (bool, [][]byte, error) {
 	log := logf.FromContext(ctx)
 	altURLs, err := cl.ListCertAlternates(ctx, certURL)
 	if err != nil {
 		return false, nil, fmt.Errorf("error listing alternate certificate URLs: %w", err)
 	}
-	// Loop over all alternative chains
-	for _, altURL := range altURLs {
-		altChain, err := cl.FetchCert(ctx, altURL, true)
+	// Loop over every chains
+	for _, chainURL := range append([]string{certURL}, altURLs...) {
+		certChain, err := cl.FetchCert(ctx, chainURL, true)
 		if err != nil {
-			return false, nil, fmt.Errorf("error fetching alternate certificate chain from %s: %w", altURL, err)
+			return false, nil, fmt.Errorf("error fetching alternate certificate chain from %s: %w", chainURL, err)
 		}
-		// Loop over each cert in this alternative chain
-		for _, altCert := range altChain {
-			cert, err := x509.ParseCertificate(altCert)
-			if err != nil {
-				return false, nil, fmt.Errorf("error parsing alternate certificate chain: %w", err)
-			}
-			log.V(logf.DebugLevel).WithValues("Issuer CN", cert.Issuer.CommonName).Info("Found alternative ACME bundle")
-			if cert.Issuer.CommonName == preferredChain {
-				// if the issuer's CN matched the preferred chain it means this bundle is
-				// signed by the requested chain
-				log.V(logf.DebugLevel).WithValues("Issuer CN", cert.Issuer.CommonName).Info("Selecting alternative ACME bundle with a matching Common Name from %s", altURL)
-				return true, altChain, nil
-			}
+		// Check topmost certificate
+		cert, err := x509.ParseCertificate(certChain[len(certChain)-1])
```

Comment:
I think it might be non-ideal to backport this change, some non-let's encrypt use cases might rely on selecting a chain using the 2nd or nth certificate.

For the patch release, we can maybe pick the chain which has a match closes to the top?

And make a breaking change in the next full cert-manager release?

@SgtCoDFish 2nd opinion would be good here.

Reply:
`Matching closest to top` might still break weird edge cases, like primary cert chain and alt cert chain having identical intermediate CA on the same depth, thus we cannot ensure 100% compatibility if we pick the closest chain.

However, if we still want to cover some edge cases like you've mentioned, there's no reason not to do it.